geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/CacheClientNotifier.java

```diff
@@ -599,6 +599,18 @@ public static void notifyClients(InternalCacheEvent event) {
     }
   }
 
+  public static boolean singletonHasClientProxies() {
+    CacheClientNotifier instance = ccnSingleton;
+    if (instance != null) {
+      return instance.hasClientProxies();
+    }
+    return false;
+  }
+
+  private boolean hasClientProxies() {
+    return !this._clientProxies.isEmpty() || !this._initClientProxies.isEmpty();
```

Comment:
I think there a possible race condition where we check the first isEmpty() method, thread is interrupted, then in the different thread we move a clientProxy from initClientProxies into the clientProxies map.  We now resume the first thread and  check initClientProxies and this is also now empty, but we really had 1 client proxy the entire time?

Reply:
I think you are right. When we promote a proxy from the init map to the clientProxies map, we add it to clientProxies first and then remove it from init. So it is in both maps for a bit.
So I think this code can be fixed by checking the init map first. Does this sound correct?